File: src/main/java/com/github/zxh/classpy/classfile/ClassParser.java
package com.github.zxh.classpy.classfile;

import com.github.zxh.classpy.classfile.constant.ConstantPool;
import com.github.zxh.classpy.common.BytesComponent;
import com.github.zxh.classpy.common.BytesParser;

public class ClassParser implements BytesParser {
    
    public ClassFile parse(byte[] bytes) {
        ClassFile cf = new ClassFile();
        cf.read(new ClassReader(bytes));
        postRead(cf, cf.getConstantPool());
        return cf;
    }

    private static void postRead(ClassComponent cc, ConstantPool cp) {
        for (BytesComponent c : cc.getComponents()) {
            postRead((ClassComponent) c, cp);
        }
        cc.postRead(cp);
    }

}

File: src/main/java/com/github/zxh/classpy/classfile/ClassParser.kt
package com.github.zxh.classpy.classfile

import com.github.zxh.classpy.classfile.constant.ConstantPool
import com.github.zxh.classpy.classfile.reader.ClassReader

object ClassParser {

    @JvmStatic
    fun parse(bytes: ByteArray): ClassFile {
        val cf = ClassFile()
        cf.read(ClassReader(bytes))
        afterRead(cf, cf.getConstantPool())
        return cf
    }

    fun afterRead(cc: ClassComponent, cp: ConstantPool) {
        cc.subComponents?.forEach { c -> afterRead(c, cp) }
        cc.afterRead(cp)
    }

}
